### PR TITLE
Limit number of checks sent over to the Link Checker API

### PR DESCRIPTION
**Pull request by SebAshton**
This PR limits the number of editions sent to be checked to 25 per organisation. It will take any new publication first, and then the editions with the oldest checks.

The idea is to monitor the number of jobs created and increase over time.

If every `Organisation` has more than 25 public editions then this limit would ensure only `25750` checks were created on the nightly schedule.

### CODE REVIEW

File: app/workers/check_organisation_links_worker.rb [hunk 1]
```diff
@@ -1,13 +1,14 @@
 # Calls the Link Checker API to verify all links in public editions, either per organisation or not
 class CheckOrganisationLinksWorker
   include Sidekiq::Worker
+  ORGANISATION_EDITION_LIMIT = 25
```
Comment (thomasleese) [inline app/workers/check_organisation_links_worker.rb]: I would definitely aim for something more like `1000` here.

File: app/workers/check_organisation_links_worker.rb [hunk 2]
```diff
@@ -1,13 +1,14 @@
...
 
   sidekiq_options queue: "link_checks"
 
   def perform(organisation_id)
     organisation = find_organisation(organisation_id)
 
-    public_editions(organisation).find_each do |edition|
+    public_editions(organisation).each do |edition|
```
Comment (thomasleese) [inline app/workers/check_organisation_links_worker.rb]: Can we not use `find_each` here?

File: test/unit/workers/check_organisation_links_worker_test.rb [hunk 3]
```diff
@@ -23,15 +23,59 @@ class CheckOrganisationLinksWorkerTest < ActiveSupport::TestCase
   end
 
   test "when given an organisation it only calls LinkCheckerApiService with editions for that organisation" do
-    stub_published_publication
+    checked_edition = stub_published_publication
```
Comment (thomasleese) [inline test/unit/workers/check_organisation_links_worker_test.rb]: Should this really be a `non_checked_edition`? As I understand it, it's a new edition that hasn't been checked yet?

File: app/workers/check_organisation_links_worker.rb (hunk 1 above)
Comment (SebAshton) [reply app/workers/check_organisation_links_worker.rb]: Sure, was just being conservative - perhaps too much 😄

File: app/workers/check_organisation_links_worker.rb (hunk 2 above)
Comment (SebAshton) [reply app/workers/check_organisation_links_worker.rb]: Under the hood `find_each` uses `find_in_batches` which ignores the `limit` 

> You can’t set the limit either, that’s used to control the the batch sizes.
https://apidock.com/rails/ActiveRecord/Batches/ClassMethods/find_in_batches

File: test/unit/workers/check_organisation_links_worker_test.rb (hunk 3 above)
Comment (SebAshton) [reply test/unit/workers/check_organisation_links_worker_test.rb]: the variable probably needs a better name :)

File: test/unit/workers/check_organisation_links_worker_test.rb [hunk 4]
```diff
@@ -23,15 +23,59 @@ class CheckOrganisationLinksWorkerTest < ActiveSupport::TestCase
...
+  test "when an organisation contains new and old editions" do
+    new_edition = stub_published_publication
+    checked_edition = create_and_stub_an_edition_with_checks(@hmrc)
+
+    assert_report_count_increased
+
+    assert_requested new_edition
+    assert_requested checked_edition
+  end
+
```
Comment (thomasleese) [inline test/unit/workers/check_organisation_links_worker_test.rb]: Extra empty line!

File: test/unit/workers/check_organisation_links_worker_test.rb [hunk 5]
```diff
@@ -23,15 +23,59 @@ class CheckOrganisationLinksWorkerTest < ActiveSupport::TestCase
...
     CheckOrganisationLinksWorker.new.perform(@hmrc.id)
 
     assert_equal 1, LinkCheckerApiReport.count
+    assert_requested new_edition
+  end
+
+  test "when an organisation contains more new editions than the limit" do
+    new_edition = stub_published_publication
+    checked_edition = create_and_stub_an_edition_with_checks(@hmrc)
+    another_checked_edition = create_and_stub_an_edition(@hmrc)
```
Comment (thomasleese) [inline test/unit/workers/check_organisation_links_worker_test.rb]: Can this be called `existing_edition`?

File: test/unit/workers/check_organisation_links_worker_test.rb [hunk 6]
```diff
@@ -23,15 +23,59 @@ class CheckOrganisationLinksWorkerTest < ActiveSupport::TestCase
...
+    assert_report_count_increased
+
+    assert_requested new_edition
+    assert_requested another_checked_edition
+    assert_not_requested checked_edition
+  end
+
+  test "when an organisation contains new and old editions" do
+    new_edition = stub_published_publication
+    checked_edition = create_and_stub_an_edition_with_checks(@hmrc)
```
Comment (thomasleese) [inline test/unit/workers/check_organisation_links_worker_test.rb]: Could this be called `existing_edition` perhaps?

**Comment by SebAshton** [comment]
Have addressed your comments @thomasleese 👍

**Comment by thomasleese** [comment]
We'll also need to re-enable the worker.

**Comment by SebAshton** [comment]
👍  Will be a follow-up PR

**Pull request merged**